### PR TITLE
add a new function contains_coordinate

### LINKED ISSUE #4726: Add a contains coordinate method to maputils
This would provide a quick way to see if a world coordinate is contained within the array bounds of a Map

### PR DESCRIPTION
**Pull request by SnehVachhani**
### Description

Fixes #4726

### CODE REVIEW

File: sunpy/map/maputils.py [hunk 1]
```diff
@@ -375,3 +375,25 @@ def on_disk_bounding_coordinates(smap):
...
+    Returns
+    -------
+    `~bool`
+        Returns `True` If coordinates are within the coordinate
+        bound of the smap and Returns `False` If coordinates are 
+        not within the coordinate bound of the smap.
+    """
+    _verify_coordinate_helioprojective(smap.coordinate_frame)
+
+    return coordinates in all_coordinates_from_map(smap)
```
Comment (ayshih) [inline sunpy/map/maputils.py:399]: This implementation would not give the correct answer if input coordinate did not correspond exactly with one of the pixels in the map, plus it's very inefficient.  You should convert the input coordinate to pixel space and check if it falls within the pixel bounds.

Comment (SnehVachhani) [reply sunpy/map/maputils.py:399]: i am new to OSC and this my first time trying it . "You should convert the input coordinate to pixel space and check if it falls within the pixel bounds" . please tell me how to convert coordinate to pixel space

Comment (dstansby) [reply sunpy/map/maputils.py:399]: You can use the `world_to_pixel` method: https://docs.sunpy.org/en/stable/api/sunpy.map.GenericMap.html?highlight=world_to_pixel#sunpy.map.GenericMap.world_to_pixel. Once you have the pixels, you can then check that they are within the limits of the map.

Comment (abhishekpandey1880) [reply sunpy/map/maputils.py:399]: Can we use this method? -> with the help of ```world_to_pixel()``` method, Find the pixel of coordinates and then find out the maximum and minimum values of edges of map by ```map_edges()``` method and check whether the coordinate pixels are within the max and min values. 
And I am having a doubt that, if the coordinate is out of bounds how will it give pixel to the method ```world_to_pixel()```, as it is not there in the map?

Comment (dstansby) [reply sunpy/map/maputils.py:399]: Shouldn't even need to use `map_edges()`, can just compare the pixel values to the size of the data array. In general a coordinate > pixel transformation doesn't depend on the pixel boundaries, so there's no need to worry about out of bounds coordinates.

File: changelog/4761.feature.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Added :func:`contains_coordinate()` which provides a quick way to see if a world coordinate is
```
Comment (Cadair) [inline changelog/4761.feature.rst:1]:
```suggestion
Added :func:`sunpy.map.contains_coordinate` which provides a quick way to see if a world coordinate is 
```

File: sunpy/map/maputils.py [hunk 3]
```diff
@@ -375,3 +375,29 @@ def on_disk_bounding_coordinates(smap):
...
+        `~sunpy.coordinates.Helioprojective`.
+
+    Returns
+    -------
+    `~bool`
+        Returns `True` If coordinates are within the coordinate
+        bound of the smap and Returns `False` If coordinates are
+        not within the coordinate bound of the smap.
+    """
+    _verify_coordinate_helioprojective(smap.coordinate_frame)
```
Comment (Cadair) [inline sunpy/map/maputils.py:398]: Is there any particular reason this utility needs the map to be in helioprojective?

File: sunpy/map/maputils.py [hunk 4]
```diff
@@ -375,3 +375,29 @@ def on_disk_bounding_coordinates(smap):
...
+    # Getting pixel indices of smap
+    xs, ys = all_pixel_indices_from_map(smap)
```
Comment (Cadair) [inline sunpy/map/maputils.py:400]: You don't need to calculate all the pixel indices, just calculate the pixel position of the input coordinates and ensure that the results are within the size of `smap.dimensions`.

File: sunpy/map/maputils.py [hunk 5]
```diff
@@ -375,3 +375,28 @@ def on_disk_bounding_coordinates(smap):
...
+    `~bool`
+        Returns `True` If coordinates are within the coordinate
+        bound of the smap and Returns `False` If coordinates are
+        not within the coordinate bound of the smap.
+    """
+    #Dimensions of smap
+    xs, ys = smap.dimensions
+    # Converting coordinates to pixels
+    xc, yc = smap.world_to_pixel(coordinates)
+    return xc<=xs and yc<=ys
```
Comment (ayshih) [inline sunpy/map/maputils.py:402]:
```suggestion
    return xc >= -0.5 and xc <= xs-0.5 and and yc >= -0.5 and yc <= ys-0.5
```
You need to check that the pixel is within the bounds of the other two edges as well, and you need to account for half pixels.  For example, if there are 1024 pixels along a particular dimension, the pixel numbers can go from 0 to 1023, but a coordinate resolves to within the map if its converted pixel value is within -0.5 to 1023.5.

File: sunpy/map/maputils.py [hunk 6]
```diff
@@ -375,3 +375,28 @@ def on_disk_bounding_coordinates(smap):
...
+def contains_coordinate(smap, coordinates):
+    """
+    Parameters
+    ----------
+    smap : `~sunpy.map.GenericMap`
+        A map in helioprojective Cartesian coordinates.
+
+    coordinates : `~astropy.coordinates.SkyCoord`, `~sunpy.coordinates.frames.Helioprojective`
+        The input coordinate. The coordinate frame must be
+        `~sunpy.coordinates.Helioprojective`.
```
Comment (ayshih) [inline sunpy/map/maputils.py:389]:
```suggestion
    smap : `~sunpy.map.GenericMap`
        The input map.

    coordinates : `~astropy.coordinates.SkyCoord`
        The input coordinate.
```

File: sunpy/map/maputils.py [hunk 7]
```diff
@@ -375,3 +375,28 @@ def on_disk_bounding_coordinates(smap):
     return SkyCoord([np.nanmin(tx), np.nanmax(tx)] * u.arcsec,
                     [np.nanmin(ty), np.nanmax(ty)] * u.arcsec,
                     frame=smap.coordinate_frame)
+
+
+def contains_coordinate(smap, coordinates):
+    """
```
Comment (ayshih) [inline sunpy/map/maputils.py:381]:
```suggestion
    """
    Checks whether a coordinate falls within the bounds of a map.
    
```

File: sunpy/map/maputils.py [hunk 8]
```diff
@@ -375,3 +375,28 @@ def on_disk_bounding_coordinates(smap):
...
+    coordinates : `~astropy.coordinates.SkyCoord`, `~sunpy.coordinates.frames.Helioprojective`
+        The input coordinate. The coordinate frame must be
+        `~sunpy.coordinates.Helioprojective`.
+
+    Returns
+    -------
+    `~bool`
+        Returns `True` If coordinates are within the coordinate
+        bound of the smap and Returns `False` If coordinates are
+        not within the coordinate bound of the smap.
```
Comment (ayshih) [inline sunpy/map/maputils.py:396]:
```suggestion
        Returns ``True`` if ``coordinates`` falls within the bounds of ``smap``.
```

File: sunpy/map/maputils.py [hunk 9]
```diff
@@ -375,3 +375,28 @@ def on_disk_bounding_coordinates(smap):
...
+    """
+    #Dimensions of smap
```
Comment (ayshih) [inline sunpy/map/maputils.py:398]:
```suggestion
    # Dimensions of smap
```

**Review by ayshih — CHANGES_REQUESTED** [review]

**Comment by nabobalis** [comment]
I think a unittest would be good to add.

**Comment by nabobalis** [comment]
Will you be able to update this @SnehVachhani?

**Comment by abhishekpandey1880** [comment]
> Will you be able to update this @SnehVachhani?

@SnehVachhani is little bit busy preparing for his exams. Should I continue this issue?
But I want to have some documentation of writing unit test in sunpy, I searched for it but was not able to find anywhere.
Please if anyone can help me out.

**Comment by nabobalis** [comment]
> > Will you be able to update this @SnehVachhani?
> 
> @SnehVachhani is little bit busy preparing for his exams. Should I continue this issue?
> But I want to have some documentation of writing unit test in sunpy, I searched for it but was not able to find anywhere.
> Please if anyone can help me out.

If you want to!

Basic documentation: https://docs.sunpy.org/en/latest/dev_guide/contents/tests.html#writing-tests

**Comment by dstansby** [comment]
👋 just wanted to check in and see if anyone is still working on this? If not I would be happy to pick it up, bit if someone is still working on it let me know and I won't take over.

**Comment by abhishekpandey1880** [comment]
> 👋 just wanted to check in and see if anyone is still working on this? If not I would be happy to pick it up, bit if someone is still working on it let me know and I won't take over.

Hey @dstansby , I and @SnehVachhani  are busy preparing for our semester exams, so we may not work on this. Our exams will get finished by 14th March, so after that we may continue.
But till then if you want you can take over it. 😊😊

File: sunpy/map/maputils.py [hunk 10]
```diff
@@ -375,3 +375,27 @@ def on_disk_bounding_coordinates(smap):
...
+    Returns
+    -------
+    `~bool`
+        Returns ``True`` if ``coordinates`` falls within the bounds of ``smap``.
+    """
+    # Dimensions of smap
+    xs, ys = smap.dimensions
+    # Converting coordinates to pixels
+    xc, yc = smap.world_to_pixel(coordinates)
+    return xc >= -0.5 and xc <= xs-0.5 and yc >= -0.5 and yc <= ys-0.5
```
Comment (ayshih) [inline sunpy/map/maputils.py:401]:
```suggestion
    return (xc >= -0.5*u.pix) * (xc <= xs - 0.5*u.pix) * (yc >= -0.5*u.pix) * (yc <= ys - 0.5*u.pix)
```
The existing code doesn't work

**Review by ayshih — CHANGES_REQUESTED** [review]
- Add `contains_coordinate()` to `maputils.py`'s `__all__`
- Add a unit test

**Comment by dstansby** [comment]
Thanks for working on this! I've updated and added tests in https://github.com/sunpy/sunpy/pull/5252 so we can get this into sunpy 3.0.

**Comment by nabobalis** [comment]
I echo the comments of dstansby, thanks again for getting the ball rolling on this.

**Pull request closed without merging**